File: 202. Happy Number.cpp
// Space: O(1)
class Solution {
public:
    bool isHappy(int n) {
        int slow = n, fast = n;
        do {
            slow = digitsSqrdSum(slow);
            fast = digitsSqrdSum(digitsSqrdSum(fast));
            if (fast == 1) return true;
        } while (slow != fast);
        
        return false;
    }
    
    int digitsSqrdSum(int n) {
        int sum = 0;
        for (int i = n; i > 0; i /= 10) sum += (i % 10) * (i % 10);
        return sum;
    }
};

// Extra space
class Solution {
public:
    bool isHappy(int n) {
        set<int> mark;
        int cur = n;
        while (mark.count(1) < 1 && mark.count(cur) < 1) {
            mark.insert(cur);
            int sum = 0;
            for (int i = cur; i > 0; i /= 10) sum += (i % 10) * (i % 10);
            cur = sum;
        }
        
        return mark.count(1) > 0;
    }
};

File: 202. Happy Number.py
# Time: O(logn)
# as finding the next value for a given number has a cost of O(logn)
# because we are processing each digit in the number, and the number of digits is given by logn
# and the total number of possible numbers we would encounter is either a constant or bounded by
# logn as well. detailed analysis:
# https://leetcode.com/problems/happy-number/solution/
# space: O(1)
class Solution:
    def isHappy(self, n: int) -> bool:
        def digitsSquaresSum(n):
            ssum = 0
            while n:
                ssum += (n % 10) ** 2
                n = int(n / 10)
            return ssum
        
        slow, fast = n, digitsSquaresSum(n)
        while slow != fast and fast != 1:
            slow = digitsSquaresSum(slow)
            fast = digitsSquaresSum(digitsSquaresSum(fast))
        return fast == 1
